mlite_rails_app/app/controllers/datasets_controller.rb

```diff
@@ -44,17 +55,136 @@ def update
+      "boolean"
+      # Check if all values are integers
+    elsif values.all? { |v| v.match?(/\A-?\d+\z/) }
+      "integer"
+      # Check if all values are floats
+    elsif values.all? { |v| v.match?(/\A-?\d+\.\d+\z/) }
+      "float"
+    else
+      "categorical"
+    end
+  end
+
+  # Helper method to check if a column contains boolean values
+  def boolean_column?(values)
+    boolean_patterns = %w[true false yes no 0 1]
+    values.all? { |v| boolean_patterns.include?(v.downcase) }
   end
+
+  # Helper method to detect if a column is categorical
+  def categorical?(values)
+    unique_values = values.uniq
+
+    # A column is categorical if it has fewer unique values than the threshold
+    unique_values.size < threshold
+  end
+
+
+
+  def calculate_metrics(csv_data)
+    metrics = {}
+
+    # Loop through each column and process numeric or categorical data
+    csv_data.headers.each do |column|
+      values = csv_data[column].compact
+
+      if numeric_column?(values)
+        # Numeric data: calculate mean, min, max, median, and std_dev
+        numeric_values = values.map(&:to_f)
+        metrics[column] = {
+          mean: (numeric_values.sum / numeric_values.size).round(2),
+          min: numeric_values.min,
+          max: numeric_values.max,
+          median: numeric_values.sort[numeric_values.size / 2],
+          std_dev: Math.sqrt(numeric_values.sum { |v| (v - numeric_values.sum / numeric_values.size)**2 } / numeric_values.size).round(2)
+        }
+      else
+        if not boolean_column?(values)
+          # Categorical metrics
+          unique_counts = values.tally
+          total_values = values.size
+          mode = unique_counts.max_by { |_, count| count }&.first
+
+          # Calculate entropy
+          entropy = -unique_counts.values.map { |count| count.to_f / total_values * Math.log2(count.to_f / total_values) }.sum.round(2)
+
+          # Calculate diversity index (Simpson's Diversity Index)
+          diversity_index = 1 - unique_counts.values.map { |count| (count.to_f / total_values)**2 }.sum.round(2)
+
+          metrics[column] = {
+            mode: mode,
+            unique_count: unique_counts.keys.size,
+            frequency_distribution: unique_counts,
+            top_categories: unique_counts.sort_by { |_, count| -count }.first(3).to_h, # Top 3 categories
+            entropy: entropy,
+            diversity_index: diversity_index,
+            category_percentages: unique_counts.transform_values { |count| (count.to_f / total_values * 100).round(2) },
+            least_frequent_category: unique_counts.min_by { |_, count| count }&.first,
+            null_count: csv_data[column].size - total_values # Count of missing/null values
+          }
+          end
+        end
+      end
+
+      metrics
+    end
+
+  # Helper method to check if a column contains only numeric data (integer or float)
+  def numeric_column?(values)
+    values.all? { |v| v.match?(/\A-?\d+(\.\d+)?\z/) }
+  end
```

Comment:
same thing here. change regex to `values.all? { |v| v.match?(/\A-?(?:\d+|\.\d+|\d+\.\d+)\z/) }` so it counts an edge case like `.4`